### ISSUE TITLE
Output is Printed without Ordering

### ISSUE THREAD

**Issue by mitnk** [xonsh/xonsh#2129]
I encountered this regularly, the output of some subprocesses should printed in order in other shells (Bash) and not in xonsh.

Take the `cargo` tool of the [Rust-lang](https://www.rust-lang.org) for example. Run `cargo run` in a hello world project will get different printing orders like the following:

```
$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/hello`
Hello, world!

$ cargo run
Hello, world!
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/hello`
```

While in Bash, `Hello world!` line will always be printed after the two debug info lines.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by scopatz** [comment]
Thanks for reporting this! It is a side-effect of the way that we read data from the std steams in chunks, rather than lines.  It it pretty annoying and should be fixed.

**Comment by scopatz** [comment]
I have spent the past couple of days on this, and I am convinced that it is impossible (more on this below). I am going to move this to a "wont-fix-for-now", unless someone else is willing to tackle it. 

**The workaround** is that you can declare any command where you need ordering to predict as false in the threadable predictors in the commands cache.

----------

**The details**

Ordering between any two separate pipes (stdout and stderr here) is fundamentally not guaranteed by the operating system. They are independent and so all reads and writes happen separately. This makes the problem of ordering stdout and stderr *technically* impossible.  

If you want the output to be order, you have redirect one stream into another - but then you don't have two streams anymore.  Additionally there is no way to insert a filter inside of a redirection, so there is no way to mark bytes in the merged stream by their source stream.

Now, I know what you are thinking. Every other shell in existence manages to pull this off. So even if it is *technically impossible*, it is *practically possible*. Can't xonsh just do what the other shells do? The answer is yes, I tried, and it doesn't work in practice. 

What other shells do is use `select()` and other low-level functions for selecting which stream currently has readable data. This can be used to read and then immediately write data from one side of the pipe to another stream. You still need to know how many bytes to read, or read in non-blocking mode. Those problems are solvable, even if they are platform-dependent. (Note that Python's `readline()` method on file objects blocks, and thus would not properly display prompts, so you have to read in chunks). Let's assume *we know* the number of bytes to read in - xonsh sets this to 1 kb currently.

Using `select()` from Python still fails to order stdout and stderr. This is because the whole process of specifying, starting, and reading from a subprocess in xonsh takes enough time that the pipes have already started to be filled. Thus all information on ordering is lost by the first time that we read from the pipe. 

Take for example the script:

```python
import sys

for i in range(5):
    omsg = "stdout " + str(i)
    emsg = "stderr " + str(i)
    print(omsg, flush=True, file=sys.stdout)
    print(emsg, flush=True, file=sys.stderr)
```

You would expect this to print alternating stdout and stderr message.  In Bash, we see this:

```console
scopatz@artemis:~/temp$ python x.xsh
stdout 0
stderr 0
stdout 1
stderr 1
stdout 2
stderr 2
stdout 3
stderr 3
stdout 4
stderr 4
```

However in xonsh we see, 

```console
scopatz@artemis ~/temp $ python x.xsh
stdout 0
stdout 1
stdout 2
stdout 3
stdout 4
stderr 0
stderr 1
stderr 2
stderr 3
stderr 4
```

Even dropping the chunksize to extreme levels like 1 byte or 8 bytes doesn't help because the buffers are already full. So technically any interweaving of the bytes is correct as long as the stdout bytes are ordered and the stderr bytes are ordered too.  What xonsh outputs and bash outputs just happen to be two of the more readable possibilities.

So why are the pipes full initially? Well this is because we start the subprocess and then read from its stdout and stderr. The amount of time between starting and reading is enough to fill more than a single line of each buffer.Other shells, like Bash, have a much lower latency and so their output appears ordered according to its write time, rather than its read time. Python is just too slow for this illusion. 

Additionally, I speculate that Bash and others might use some kernel properties to prevent buffers from being written to before they are ready to be read from. However, pausing pipes like this is not something that is likely cross platform and so it is not an option really open to xonsh.

So the solutions are:

1. force the output to be ordered by redirecting stderr to stdout, or stdout to stderr
2. Use Popen (rather than PopenThread) for a command to force stdout and stderr to be directly connected to the terminal emulator. Note that the prevents output from being captured.

Possible future solutions are:

3. Maybe starting up our parallel reading code before the subprocess is started would help, but I am not convinced that this would work.
4. Slug might have abstractions for dealing with this, but I suspect we'll run into the same problems.

So in summary, I think that there are two modes: 

1. One where users use Popen and ordering is promised, but output is not captured.
2. One where PopenThread is used, ordering is not promised, and output is captured.

These are the basics tradeoffs. Users have to decide for themselves what features they need. I think our stock settings are probably correct for most users in most siutations. But we could make it easier for people to pick other options.  I am very open to suggestions here!

Thanks again for reporting this @mitnk. This is a tricky one :)

**Comment by astronouth7303** [comment]
So xonsh has this problem because of one feature: Having stdout and stderr of processes as well as printing them out. No other shell does this.

For most shells, what is given as file descriptors 1 and 2 (stdout and stderr) is the tty directly. And since stdout/stderr write to the same underlying device, ordering is preserved in the stream. The shell is not involved in the process.

We can avoid the problem by redefining the API to return terminal output instead stdout, but then stderr must be redirected away or will be mixed in.

Slug will have primitives that will let us redirect and manipulate output in new and exciting ways, but there's one thing that holds true: data in two different pipes can never be correlated. If you want correlation of two streams, they must be the same pipe at the source. To change this would require kernel-level work.

**Comment by melund** [comment]
@scopatz @astronouth7303 ... Thanks for the detailed explanation. I had a lot of 'aha moments' reading this :) Could you consider an advanced section on xon.sh, where we could collect this type of content. 

Digging it out of old issue is sometimes hard.

**Comment by astronouth7303** [comment]
Maybe? There's a lot of little behaviors in POSIX that aren't really documented anywhere. I'm not sure where to start on that.

It could also go in the slug wiki, or maybe in a revamped developer's section.

**Comment by scopatz** [comment]
@astronouth7303 is right about the direct TTY connection. 

@melund That is a great idea! I'll clean it up and put in a doc PR. I think this information could go in a lot of places, but it makes a lot of sense to have something in user facing docs to me. I think that we'll get questions from users about this occasionally, and it would be good to have a place to point them. And we should implement #2119 so that the choice between the two something that has a public interface, rather than relying on `__xonsh_commands_cache__`.  It would be nice to have names for the two different modes too. Maybe `direct` and `indirect`?

**Comment by mitnk** [comment]
Hi @scopatz, thanks a lot for putting the detailed & amazing context here. It made several things clear to me. I didn't realize the output are belonging to two FDs. The explanations make a lot of sense.

I'm totally ok to mark this issue a `wont-fix-for-now`. I was just reporting an issue I notice, which is not annoying me that much.

> Other shells, like Bash, have a much lower latency and so their output appears
> ordered according to its write time, rather than its read time ...
> Additionally, I speculate that Bash and others might use some kernel properties to
> prevent buffers from being written to before they are ready to be read from.

Not 100% sure, but Bash seems using a dedicated pipe to synchronize the processes creation phase of a pipeline, so that it makes assigning controlling terminal safer and maybe also help to sync the sdtout/sdterr reading thing?

If it's true, then maybe [Slug](https://github.com/xonsh/slug) could do some similar things on this.

**Comment by astronouth7303** [comment]
> Not 100% sure, but Bash seems using a dedicated pipe to synchronize the processes creation phase of a pipeline, so that it makes assigning controlling terminal safer and maybe also help to sync the sdtout/sdterr reading thing?

Keep in mind that under most shells (ie, not xonsh), when not redirecting, the stdout and stderr are writing to the same device and are therefore the same buffers. I'm pretty sure being able to flag the source of data (stdout/stderr) while maintaining the ordering between them is something only extreme optimists can do. 

However, creating a process, assigning it to a process group, and marking that group as foreground in a tty is not atomic, meaning that it is theoretically possible for a process to write to its controlling terminal before its marked as the foreground process group. This means that it'll receive a `SIGTTOU` signal, pausing the process. (Maybe, see [A Deep Dive into the SIGTTIN / SIGTTOU Terminal Access Control Mechanism in Linux](http://curiousthing.org/sigttin-sigttou-deep-dive-linux))

This issue can't be fixed with xonsh's current capabilities. We can't accurately keep multiple streams ordered with respect to each other (with special kernel modifications), so we either have to change user expectations (xonsh will never be ordered) or what's recorded (xonsh captures what is printed to screen, not what's on stdout).

**Comment by mitnk** [comment]
> However, creating a process, assigning it to a process group, and marking that group as foreground in a tty is not atomic, meaning that it is theoretically possible for a process to write to its controlling terminal before its marked as the foreground process group.

No, it's not atomic, that's why Bash using a pipe to sync them, to make sure process run (by calling execxx) after all processes of the foreground process group fully created.

**Comment by astronouth7303** [comment]
@mitnk Can you find the source for that? My experience with POSIX so far are that shells are pretty simple things; it's the kernel that does all the hard work.

**Comment by mitnk** [comment]
Source of which part?

**Comment by astronouth7303** [comment]
The bit where bash has a syncronization pipe.

**Comment by mitnk** [comment]
It's in bash's source code (I was checking with version 3.2.57), file `jobs.c`. The pipe is this one:
```
/* Pipes which each shell uses to communicate with the process group leader
   until all of the processes in a pipeline have been started.  Then the
   process leader is allowed to continue. */
int pgrp_pipe[2] = { -1, -1 };
```

**Comment by astronouth7303** [comment]
Oh sheesh. I'm going to have to tear that part at a future point when I have brain power.

**Comment by astronouth7303** [comment]
(To be clear, coordination of processes, proc-groups, and foregrounding is a different problem than trying to coordinate distinct pipes.)

**Comment by jnoortheen** [comment]
This is no longer the case when output capturing is disabled starting from `0.11`
